### ISSUE TITLE
Crash on startup when using NVIDIA GTX 1070 (audio-related)

### ISSUE THREAD

**Issue by signiss7** [hrydgard/ppsspp#12845]
I have installed PPSSPP for the first time today. I pulled version 1.9.3 directly from the ppsspp.org page. I am using the Windows 10 version. I have tried the zip file version and regular installer. My OS is Windows 10 1909 and is up to date. It is also the 64 bit version of Windows 10. I am running an i7 770k and have a GTX 1070 TI with the latest NVidia drivers as of today. 

My PPSSPP will sometimes launch perfectly and be able to run any game I have installed. However, more frequently it crashes before it fully boots. I get two different errors when it crashes (I have 2 screenshots below of them). The first is the "defragment must not run on the render thread." The second, which is less common is, "Failed initializing any graphics. Try upgrading your graphics driver." Both of these errors occur regardless of what my graphics settings are in the application itself. However, the few times I have gotten it to boot, I have followed the troubleshooting guide on GitHub including adjusting system and graphics settings. I would list the games I am attempting, but the crashes occur before I can even launch them in nearly every case. I have also tried as many older versions as I could get a hold of. This includes as far back as version 1.5.4. 

I have attempted both 32 bit and 64 bit just to see if there was an issue there. Sometimes it will launch and let me run everything, sometimes it crashes as soon as I click on anything in the application, sometimes it will work fine for a few minutes and then crash, and sometimes it will crash on launch before I can click on anything. The debugger and logs have not been able to pickup anything as far as I can tell, as it usually crashes before I can open either. I have also rebooted to verify it was not a bad boot on my part. I attempted to reinstall the program multiple times with a clean download of the program. I also tried installing it on different SSDs, and different folder locations. Lastly, I have my graphics settings in a screenshot to show it does see my GPU, and does sometimes  work.

![ppsspp graphics error](https://user-images.githubusercontent.com/64050983/79814204-ab6c2a80-834b-11ea-9447-e19649df557c.png)

![ppsspp defargment render error](https://user-images.githubusercontent.com/64050983/79814218-b0c97500-834b-11ea-96fd-f08c8ebe9b05.png)

![ppsspp graphics settings](https://user-images.githubusercontent.com/64050983/79814235-b9ba4680-834b-11ea-91dc-c00bf2af9372.png)

**Comment by unknownbrackets** [comment]
Try the latest git build:
https://buildbot.orphis.net/ppsspp/

The defragment error has been fixed, but only occurs when PPSSPP crashes while setting up graphics with your graphics driver.  This probably means something is causing that to be unstable.  It can also mean that you're using modding software (such as GeDoSaTo, etc.) that are crashing or causing crashes.

-[Unknown]

**Comment by signiss7** [comment]
Thank you so much for the quick reply. I removed the version I had and installed the latest Win 10 version on the latest git build.  I now get a different error nearly every time I launch it. I do not have any modding software that I know of. Do you have any ideas what would cause it to be unstable?


![Annotation 2020-04-20 222742](https://user-images.githubusercontent.com/64050983/79818835-b5dff180-8356-11ea-8563-90c0a96779b2.png)

**Comment by unknownbrackets** [comment]
Strange.  A few additional questions:

Is it always switching from Direct3D 11 to Vulkan, or does it bounce around?

Do saves work?  Do settings persist?  Or are you losing them when it fails to start?

I'm assuming you sometimes get it crashing multiple times in a row.  If you press the Windows key + Ctrl + Shift + B (this shortcut restarts your graphics driver), does it temporarily stop crashing?  It would probably show at least one more "PPSSPP crashed while initializing graphics."

I'm assuming you've played other 3D games on this computer without seeing any issues with unexpected crashes?  Have you tried other emulators like Dolphin or PCSX2?

-[Unknown]

**Comment by signiss7** [comment]
It bounces around. I have not gotten far enough into anything to try a save. Settings don't appear to persist. It looks like it loses what I change in the settings. It does crash multiple times. I have also noticed that the Github version does not show up as a program in my list of programs on control panel like the other versions I tried did. Resetting the graphics driver as you suggested prompted the same error. Except this time I hit ok and it went to the application menu, but it then hung without giving an error. I had to close it via task manager. I regularly play other 3D games on this PC but have not tried other emulators on this build. Temperature looks fine. I am not overclocking anything, but I think my 1070 TI is the superclocked version of the card.

**Comment by hrydgard** [comment]
Extremely odd given how well it generally works on any Windows computer an NVIDIA card. I can't really think of anything, maybe you have some broken installation of some video capture software like Bandicam? No clue...

**Comment by unknownbrackets** [comment]
If settings aren't persisting, maybe Direct3D 11 is always crashing, and it keeps reverting to defaults.

Some things that can cause settings to revert:
1. Running PPSSPP from within a compressed folder.  Instead, extract to a folder and run from there.
2. Permission issues.  If the "memstick" folder next to PPSSPP is not writable, and the PPSSPP folder in My Documents is also not writable (for example, because they are owned by Administrator), PPSSPP won't be able to save settings.  Try deleting the "memstick" folder and make sure you can create new folders without providing a UAC approval.
3. Compatibility mode.  If PPSSPP is being run in a "compatibility mode" for an older Windows version, it might cause odd behavior.  If you right click -> Properties, you can make sure it's not running in this way.  It shouldn't by default, though.
4. Some sort of virus scanner or other software intercepting the writes.  But I don't think this is happening?

Yeah, it also works fine on my NVIDIA card...

I heard some user mentioned there was a feature that caused PPSSPP to crash, I think it was some upscaling setting (off by default) in the driver settings.  Another user reported having to install the Vulkan part of the driver when updating.

We've also had people report issues with Windows' game mode, when it does automatic screen recording.

The git build, from the zip, doesn't "install".  That's why it doesn't show in your programs list.

-[Unknown]

**Comment by signiss7** [comment]
Upon further testing, I can get both Dolphin and PCSX2 working without any issues. Am I missing something obvious about setting up PPSSPP?

**Comment by hrydgard** [comment]
I don't see what else you could have missed. This simply should work... :/

**Comment by LunaMoo** [comment]
Did you ever used something like Nvidia Inspector or manually changed nvidia settings? If yes, maybe whatever you changed is the culprit there.

 Also did you tried to reinstall GPU drivers? That is "perform a clean install" by allowing nvidia installer to uninstall old drivers first, sometimes weird things happen with old drivers leftovers.

 Any chance you had different GPU there previously?
 I recall someone had a lot of problems by changing GPU from one side to another(AMD/NVIDIA) where his old gpu drivers were still there causing crashes with the new one. It's generally good idea to completely purge old drivers for hardware you're replacing with something else.

**Comment by unknownbrackets** [comment]
These are all good ideas, but it's honestly confounding that this would ONLY affect PPSSPP.

It's true that GPU drivers frequently target specific games, software, etc. with custom driver code.  No idea if PPSSPP is such a target, but it seems unlikely (?) since it's mostly for marketing/partnership/etc. purposes AFAIK.

Settings not persisting might just be from crashing before it can save them.  But if it's not, that still seems like a key part of the issue.

If you have the latest git build of PPSSPP extracted in a folder like so:

```
ppsspp/
  PPSSPPWindows64.exe
  assets/
    ...
  memstick/
    PSP/
      SYSTEM/
        ppsspp.ini
```

Go into the "memstick" folder next to PPSSPPWindows64.exe, then PSP -> SYSTEM.  If they don't exist, create them so they are one inside another.

Now, if ppsspp.ini exists, open it and delete everything inside it.  If it doesn't, open Notepad and save a blank file to this folder.  When saving, put `"ppsspp.ini"` (with the quotes) as the filename.  Without those quotes, it might actually name the file "ppsspp.ini.txt" which we don't want.

Now, replace the contents of the file with:

```
[Graphics]
GraphicsBackend = 0 (OPENGL)
FailedGraphicsBackends = IGNORE
```

And try starting PPSSPP again.  If this doesn't work, let's try once more:

```
[Graphics]
GraphicsBackend = 1 (DIRECT3D9)
FailedGraphicsBackends = IGNORE
```

These will force a specific backend setting.  I'm wondering if Direct3D 11 specifically is crashing for you, and something is going wrong with changing the settings.

If nothing changes at all (including it saying DIRECT3D11 -> still sometimes), then something has gone wrong with forcing the settings.  Maybe PPSSPP is not reading the ini you saved.

-[Unknown]

**Comment by signiss7** [comment]
For LunaMoo's questions, old drivers are purged. It is just the 1070ti drivers, but really good guess. As for unknownbrackets, I tried each of your steps and it will not launch at all. 

I then removed and reinstalled the application again with today's 64 bit version on Github rather than the one from the other day. I tried both steps again. It appears that OpenGL failed immediately but Direct3D9 worked fine on first boot. I clicked through some settings for a bit just to see if it would crash. After that, I closed it and attempted a second boot. It failed the second, third, and fourth boot. The application starts to boot up, enough that I can see the setting to stay on Direct 3D9 is staying on each attempted boot. I tried OpenGL again just to see if it would stick no matter what I put. It does stick with whatever I set the .ini file to. It just does not boot after the first boot.

**Comment by unknownbrackets** [comment]
Well, that's an interesting result.  So what we know:

 * It can run fine, but only once after deleting.
 * When it crashes, it may happen immediately at start, or after some graphics (like the logo) have rendered.
 * Nothing is making it run in compatibility mode after the first run.
 * Even resetting ini settings after running doesn't prevent the crash.
 * Direct3D 9 might slightly more stable but it's possible that's happenstance.

What if you delete the "memstick" folder and then try the ppsspp.ini thing again with DIRECT3D9?  There are other files it creates in memstick, so I just want to confirm if these are what's different between the first time and the times it crashes.

Are you using any controllers or gamepads?  Is it possible to temporarily try unplugging them to see if there's any change?  PPSSPP checks for gamepad changes periodically and on startup.  It's caused crashes before for people with a USB controller problem.  I thought of this when considering how it may crash later, but it doesn't really explain it working that one time.  We might be checking differently than Dolphin/PCSX2, though.

What happens if you rename the exe?  Assuming none of the above helped, try renaming "PPSSPPWindows64.exe" to "Testing.exe".  Alternatively, move it to a new folder.  Does it work once then?  If this helps, it could mean many things, including things Windows remembers between runs.  That would largely eliminate any consideration of a hardware issue, it'd have to be at the software level.

My goal is to reproduce the ability to run it successfully once with the least changes.  That'll narrow down what's going wrong to whatever we changed to make it (temporarily) work.

If none of that helps, can you try a super old version of PPSSPP?  This is pretty old:
http://ppsspp.org/files/1_0_1/ppsspp_win.zip

That's before we made a bunch of changes and possibly more similar to Dolphin.  It won't run games as well and is lacking many features, but if that runs consistently it'd tell us a lot.  If that also fails in exactly the same way, it'd be very interesting too.

-[Unknown]

**Comment by unknownbrackets** [comment]
Just a note for posterity: a GT 1030 and a GTX 1070M were both tested, on the latest drivers, without any of these crashes or issues.

-[Unknown]

**Comment by Kondraki643** [comment]
Hello! I'm having the same issue on my GTX 1660 Super. I've tried: 
- Reverting monitor settings in NVidia control panel
- Installing the latest git build and editing ppsspp.ini

What actually worked:
- Launching a really old version seems running fine, at least on first boot

**Comment by unknownbrackets** [comment]
Can you try newer versions?  It'd be extremely helpful if you can find the first git build that stopped working.

If you go here:
https://buildbot.orphis.net/ppsspp/index.php?m=fulllist

You'll see we have all the git builds since v1.6.3.  Don't worry, testing every single one would be crazy.  There are 2577 changes since then, but if you test a build in the middle, you'll eliminate half of those in one shot.

So if you could test v1.8.0-96-gb5db387bf, that's about halfway between the two.  If it works, the problem is more recent - if it doesn't work, it's likely earlier.

With one more try, you could cut away another 600 possible causes.  In fact, if you tried 11 different builds you could narrow it down pretty precisely to a single change.  On the other hand, since I can't reproduce the issue - I'm left staring at a haystack of 2577 commits, wondering where the needle might be.  You're the hero here who can cut down those commits like the small fry they are.

If v1.8.0-96-gb5db387bf also fails, the next one half way down is v1.7.1-30-g886a1e031.
If v1.8.0-96-gb5db387bf works like v1.6.3 did, half way up would be about v1.9.3-51-g56e2e0845.

The number after the first dash is the number of changes since that stable version, so it gets easier to calculate after that.  If you could test 3 builds, or especially a few more than that, it would give us a decent shot at identifying what change makes this problem happen.

-[Unknown]

**Comment by unknownbrackets** [comment]
Another person on Discord had a problem that seemed similar to this.

They followed the above steps and it seems like #11438 was where the problem started happening.  That's when we started detecting audio device changes (such as plugging in a headset, or switching your audio output to your TV.)

This could indicate a bug in the WASAPI device switching/detection, or it could indicate a driver/hardware problem related to audio.

A note that NVIDIA is at least related to audio.  For example, if you use HDMI and your monitor/TV outputs sound, the NVIDIA card is handling that.  Maybe something about how we're detecting audio device changes is making the video side of things fail to start, or maybe we're simply crashing during audio init and mistaking that for a graphics init crash.

-[Unknown]

**Comment by unknownbrackets** [comment]
The latest build should have more safety checks in the audio code.

If it still crashes, please try searching for this in ppsspp.ini:
```
AudioBackend = 0
```

And change it to:
```
AudioBackend = 1
```

This will move you to the legacy audio backend.  If that stops it from crashing, then it's not a graphics problem at all.

-[Unknown]

**Comment by hrydgard** [comment]
I don't know how to confirm whether we've fixed this or not.. We've definitely fixed a number of plausible causes, whether that translates to this no longer happening in the wild is unclear. 

Thinking of bumping further action to 1.11.

**Comment by unknownbrackets** [comment]
I will say that we haven't heard any reports of this happening for a while.  Not sure if people are finding this and switching to DirectSound or using an ancient build, but people aren't noticing this problem in Discord either.  Maybe a driver bug got fixed.

-[Unknown]

**Comment by unknownbrackets** [comment]
Closing this because I suspect it's been fixed (possibly outside PPSSPP.)

It's hard to tell when fixes in PPSSPP might fix other issues, and sometimes certain drivers may cause bugs that can't be fixed.

If you have more information or can confirm it still happens in the latest git builds, please reply to this issue and it'll be reopened.  If you have a new issue with the same game, just create a new issue instead.

-[Unknown]
